presto-native-execution/presto_cpp/main/common/prometheus-metrics/PrometheusReporter.cpp

```diff
@@ -0,0 +1,172 @@
+  }
+  int numBuckets = (max - min) / bucketWidth;
+  auto bound = min + bucketWidth;
+  std::string keyStr = std::string(key);
+  std::replace(keyStr.begin(), keyStr.end(), '.', '_');
+  auto& histogramFamily =
+      ::prometheus::BuildHistogram().Name(keyStr).Register(*registry_);
+  ::prometheus::Histogram::BucketBoundaries bucketBoundaries;
+  while (numBuckets) {
+    bucketBoundaries.push_back(bound);
+    bound += bucketWidth;
+    numBuckets--;
+  }
+  VELOX_CHECK_GE(bucketBoundaries.size(), 1);
+  auto& histogramMetric = histogramFamily.Add(labels_, bucketBoundaries);
+  histogramMap_.emplace(key, histogramMetric);
+  // If percentiles are provided, create a Summary type metric and register.
+  if (pcts.size() > 0) {
+    auto& summaryFamily = ::prometheus::BuildSummary()
+                              .Name(keyStr.append(kSummarySuffix))
+                              .Register(*registry_);
+    ::prometheus::Summary::Quantiles quantiles;
+    for (auto pct : pcts) {
+      quantiles.push_back(
+          ::prometheus::detail::CKMSQuantiles::Quantile(pct / (double)100, 0));
+    }
+    auto& summaryMetric = summaryFamily.Add({labels_}, quantiles);
+    summaryMap_.emplace(std::string(key), summaryMetric);
+  }
+  registeredHistogramMetrics_.insert(key);
+}
+
+void PrometheusReporter::registerHistogramMetricExportType(
+    folly::StringPiece key,
+    int64_t bucketWidth,
+    int64_t min,
+    int64_t max,
+    const std::vector<int32_t>& pcts) const {
+  registerHistogramMetricExportType(key.begin(), bucketWidth, min, max, pcts);
+}
+
+void PrometheusReporter::addMetricValue(const std::string& key, size_t value)
+    const {
+  addMetricValue(key.c_str(), value);
+}
+
+void PrometheusReporter::addMetricValue(const char* key, size_t value) const {
+  if (!registeredMetrics_.count(key)) {
+    VLOG(1) << "addMetricValue for unregistred metric " << key;
+    return;
+  }
+  auto statType = registeredMetrics_.find(key)->second;
+  std::lock_guard<std::mutex> l(mutex_);
+  switch (statType) {
+    case velox::StatType::COUNT:
+      countersMap_.find(key)->second.Increment(value);
+      break;
+    case velox::StatType::SUM:
+    case velox::StatType::AVG:
+    case velox::StatType::RATE:
+      // Overrides the existing state.
+      gaugeMap_.find(key)->second.Set(value);
+      break;
+  };
+}
+
+void PrometheusReporter::addMetricValue(folly::StringPiece key, size_t value)
+    const {
+  addMetricValue(key.begin(), value);
+}
+
+void PrometheusReporter::addHistogramMetricValue(
+    const std::string& key,
+    size_t value) const {
+  addHistogramMetricValue(key.c_str(), value);
+}
+
+void PrometheusReporter::addHistogramMetricValue(const char* key, size_t value)
+    const {
+  if (!registeredHistogramMetrics_.count(key)) {
+    VLOG(1) << "addMetricValue for unregistered metric " << key;
+    return;
+  }
+  std::lock_guard<std::mutex> l(mutex_);
+  histogramMap_.find(key)->second.Observe(value);
+  if (summaryMap_.count(key)) {
```

Comment:
Since you added a suffix to the summaryMap when registering, this key will never exist

Reply:
The summaryMap uses key, but for registeredHistogramMetrics_ of histograms I am adding a suffix to uniquely identify a summary type.